File: AutomationService/AutomationService/Features/Vorgaenge/Domain/Services/RegisterSpiegelErgebnis.cs
namespace AutomationService.Features.Vorgaenge.Domain.Services;

/// <summary>
/// Was beim Schreiben des Register-Spiegels herauskam (§6.2, #40).
///
/// Ein Ergebnis statt einer Ausnahme, weil der häufigste Fehlerfall — das
/// Register ist gerade in Word geöffnet — kein Programmfehler ist, sondern
/// eine Lage, die die Oberfläche erklären und der Anwender in fünf Sekunden
/// beheben kann. Vor allem aber darf er den Vorgangsabschluss nicht mitreißen,
/// nach dem der Spiegel läuft.
/// </summary>
/// <param name="Geschrieben">Ob in diesem Lauf Dateien entstanden sind.</param>
/// <param name="Grund">
/// Warum nicht geschrieben wurde — kein Ablageordner eingestellt, oder der
/// Bestand ist unverändert. Null, wenn geschrieben wurde.
/// </param>
/// <param name="Fehler">
/// Klartext für die Oberfläche, wenn das Schreiben scheiterte. Null sonst.
/// </param>
/// <param name="DocxPfad">Die geschriebene Word-Datei; null, wenn keine entstand.</param>
/// <param name="PdfPfad">
/// Die abgelegte PDF-Datei; null, wenn keine liegt. Direkt nach einem
/// Schreiblauf ist das der Normalfall — die Fassung entsteht erst danach
/// (§6.2 „Word sofort, PDF nachgezogen"), und was dann zählt, ist
/// <see cref="PdfLaeuft"/>.
/// </param>
/// <param name="PdfFehler">
/// Warum das PDF fehlt, obwohl die .docx steht — auf einem Rechner ohne Word
/// ist das ein erwartbarer Zustand und kein Grund, den Spiegel als
/// gescheitert zu melden. Null, solange die Fassung noch entsteht: „noch nicht
/// da" ist kein Fehler und darf nicht als einer erscheinen.
/// </param>
/// <param name="PdfLaeuft">
/// Ob gerade eine PDF-Fassung entsteht (§6.2: „Dass ein PDF gerade entsteht,
/// ist an der Oberfläche ablesbar"). Gilt für den ganzen Prozess und nicht nur
/// für diesen Aufruf — wer die Registerseite nach einem Fensterwechsel neu
/// öffnet, sieht den laufenden Lauf weiterhin.
/// </param>
/// <param name="Zeilen">Anzahl der Registerzeilen in der Datei.</param>
/// <param name="GeschriebenAm">Zeitpunkt des letzten erfolgreichen Schreibens.</param>
/// <param name="Konfliktkopien">
/// Dateien im Ablageordner, die nach einer vom Synchronisierungsdienst
/// angelegten Konfliktkopie aussehen. Ihr Auftauchen heißt: Jemand hat den
/// Spiegel unterwegs bearbeitet — ab da gäbe es zwei Register, und genau davor
/// soll die Oberfläche warnen.
/// </param>
public sealed record RegisterSpiegelErgebnis(
    bool Geschrieben,
    string? Grund,
    string? Fehler,
    string? DocxPfad,
    string? PdfPfad,
    string? PdfFehler,
    bool PdfLaeuft,
    int Zeilen,
    DateTime? GeschriebenAm,
    IReadOnlyList<string> Konfliktkopien)
{
    /// <summary>
    /// Es gab nichts zu tun — kein Ablageordner eingestellt, oder der Bestand
    /// ist unverändert.
    ///
    /// <see cref="Konfliktkopien"/> wird durchgereicht und nicht weggelassen:
    /// Eine Warnung, die nur auf dem Erfolgsweg erscheint, fehlt gerade dann,
    /// wenn der Anwender sie braucht — ein übersprungener Lauf ist der
    /// Normalfall, und die Konfliktkopie daneben verschwindet dadurch nicht.
    ///
    /// <see cref="PdfLaeuft"/> bleibt hier <c>false</c> und wird erst am
    /// Ausgang des Dienstes gesetzt — an einer Stelle, aus einer Quelle
    /// (siehe <c>RegisterSpiegelService</c>).
    /// </summary>
    public static RegisterSpiegelErgebnis Uebersprungen(
        string grund, int zeilen, DateTime? zuletzt, IReadOnlyList<string>? konfliktkopien = null) =>
        new(false, grund, null, null, null, null, false, zeilen, zuletzt, konfliktkopien ?? []);

    /// <summary>
    /// Das Schreiben scheiterte an einer Lage, die der Anwender beheben kann.
    /// <see cref="Konfliktkopien"/> wird aus demselben Grund durchgereicht wie
    /// bei <see cref="Uebersprungen"/>, <see cref="PdfLaeuft"/> aus demselben
    /// Grund nicht gesetzt.
    /// </summary>
    public static RegisterSpiegelErgebnis Gescheitert(
        string fehler, int zeilen, DateTime? zuletzt, IReadOnlyList<string>? konfliktkopien = null) =>
        new(false, null, fehler, null, null, null, false, zeilen, zuletzt, konfliktkopien ?? []);
}

File: Automation_App_Frontend/lib/features/vorgaenge/domain/entities/register_spiegel_ergebnis.dart
import 'package:equatable/equatable.dart';

/// Was der Register-Spiegel zuletzt getan hat (§6.2) — das Gegenstück zu
/// `RegisterSpiegelErgebnis`/`RegisterSpiegelDto` im Backend.
///
/// Bewusst nicht „Stand": So heißt im Backend die *Merkdatei*, die verhindert,
/// dass ein unveränderter Bestand zweimal geschrieben wird. Zwei Klassen
/// gleichen Namens auf beiden Seiten desselben Vertrags, die Verschiedenes
/// meinen, sind genau die Falle, in die eine Suche später läuft.
///
/// Der Spiegel ist die Word-/PDF-Fassung des Registers in einem einstellbaren
/// Ordner — liegt der im synchronisierten Bereich (OneDrive), ist das Register
/// unterwegs lesbar, ohne dass die App etwas von der Cloud wissen muss.
/// Geschrieben wird ausschließlich in der App; die Datei dort ist ein Spiegel
/// und kein zweites Original.
class RegisterSpiegelErgebnis extends Equatable {
  /// Ob in diesem Lauf Dateien entstanden sind.
  final bool geschrieben;

  /// Warum nicht geschrieben wurde — kein Ablageordner eingestellt, oder der
  /// Bestand ist unverändert. Null, wenn geschrieben wurde.
  final String? grund;

  /// Klartext, wenn das Schreiben scheiterte (fast immer: Die Datei ist noch
  /// in Word geöffnet). Null sonst.
  final String? fehler;

  final String? docxPfad;
  final String? pdfPfad;

  /// Warum das PDF fehlt, obwohl die Word-Datei steht. Auf einem Rechner ohne
  /// Word ist das erwartbar und kein Fehlschlag: Die `.docx` ist die
  /// verbindliche Fassung, das PDF die bequeme.
  final String? pdfFehler;

  /// Ob gerade eine PDF-Fassung entsteht (§6.2 „Word sofort, PDF
  /// nachgezogen"). Direkt nach dem Export gilt regelmäßig `pdfPfad: null`,
  /// `pdfFehler: null`, `pdfLaeuft: true` — „noch nicht da" ist ausdrücklich
  /// **kein** Fehler und darf nicht als solcher gemeldet werden (§6.2
  /// „Solange kein neues PDF liegt, liegt auch kein altes"). Fertig meldet der
  /// Hub `/hubs/register` mit `registerPdfFertig`
  /// ([RegisterPushNotifier.onPdfFertig]); danach steht hier wieder `false`.
  final bool pdfLaeuft;

  final int zeilen;
  final DateTime? geschriebenAm;

  /// Dateien neben dem Spiegel, die nach einer Konfliktkopie des
  /// Synchronisierungsdienstes aussehen. Ihr Auftauchen heißt: Jemand hat den
  /// Spiegel unterwegs bearbeitet — ab da gäbe es zwei Register.
  final List<String> konfliktkopien;

  const RegisterSpiegelErgebnis({
    this.geschrieben = false,
    this.grund,
    this.fehler,
    this.docxPfad,
    this.pdfPfad,
    this.pdfFehler,
    this.pdfLaeuft = false,
    this.zeilen = 0,
    this.geschriebenAm,
    this.konfliktkopien = const [],
  });

  /// Der Zustand vor dem ersten Abruf.
  static const RegisterSpiegelErgebnis unbekannt = RegisterSpiegelErgebnis();

  factory RegisterSpiegelErgebnis.fromJson(Map<String, dynamic> json) {
    final zeitpunkt = json['geschriebenAm'] as String?;
    return RegisterSpiegelErgebnis(
      geschrieben: json['geschrieben'] as bool? ?? false,
      grund: json['grund'] as String?,
      fehler: json['fehler'] as String?,
      docxPfad: json['docxPfad'] as String?,
      pdfPfad: json['pdfPfad'] as String?,
      pdfFehler: json['pdfFehler'] as String?,
      pdfLaeuft: json['pdfLaeuft'] as bool? ?? false,
      zeilen: (json['zeilen'] as num?)?.toInt() ?? 0,
      // `toLocal()`: der Dienst sendet mit Zeitzonenversatz, angezeigt wird
      // Ortszeit. Ohne das nennt die Fußleiste zwei Stunden früher als das
      // geschriebene Dokument selbst — und lässt den Schreibvorgang wie
      // ausgefallen aussehen.
      geschriebenAm: zeitpunkt == null
          ? null
          : DateTime.tryParse(zeitpunkt)?.toLocal(),
      konfliktkopien:
          (json['konfliktkopien'] as List?)?.cast<String>() ?? const [],
    );
  }

  /// Für [RegisterSpiegelCubit._pdfNachgezogen]: Trägt die Hub-Meldung nach,
  /// ohne den Rest des zuletzt geladenen Stands anzurühren.
  RegisterSpiegelErgebnis copyWith({
    bool? geschrieben,
    String? grund,
    String? fehler,
    String? docxPfad,
    String? pdfPfad,
    String? pdfFehler,
    bool? pdfLaeuft,
    int? zeilen,
    DateTime? geschriebenAm,
    List<String>? konfliktkopien,
  }) => RegisterSpiegelErgebnis(
    geschrieben: geschrieben ?? this.geschrieben,
    grund: grund ?? this.grund,
    fehler: fehler ?? this.fehler,
    docxPfad: docxPfad ?? this.docxPfad,
    pdfPfad: pdfPfad ?? this.pdfPfad,
    pdfFehler: pdfFehler ?? this.pdfFehler,
    pdfLaeuft: pdfLaeuft ?? this.pdfLaeuft,
    zeilen: zeilen ?? this.zeilen,
    geschriebenAm: geschriebenAm ?? this.geschriebenAm,
    konfliktkopien: konfliktkopien ?? this.konfliktkopien,
  );

  @override
  List<Object?> get props => [
    geschrieben,
    grund,
    fehler,
    docxPfad,
    pdfPfad,
    pdfFehler,
    pdfLaeuft,
    zeilen,
    geschriebenAm,
    konfliktkopien,
  ];
}
